3p/iframe-messaging-client.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import './polyfills';
+import {listen} from '../src/event-helper';
+import {getRandom} from '../src/3p-frame';
+import {user} from '../src/log';
+/**
+ * @abstract
+ */
+export class IframeMessagingClient {
+
+  /**
+   *  @param {Window} win A window object.
+   */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+    /** Map messageType keys to callback functions for when we receive
+     *  that message
+     *  @private {object}
+     */
+    this.callbackFor_ = {};
+    this.setupEventListener_();
+  }
+
+  /**
+   * Register callback function for message with type messageType.
+   *   As it stands right now, only one callback can exist at a time.
+   *   All future calls will overwrite any previously registered
+   *   callbacks.
+   * @param {string} messageType The type of the message.
+   * @param {function(object)} callback The callback function to call
+   *   when a message with type messageType is received.
+   */
+  registerCallback_(messageType, callback) {
```

Comment:
there is already a listenFor function in use, would rather be able to disambiguate